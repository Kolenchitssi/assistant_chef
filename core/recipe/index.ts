export * from './recipe.models';
export {
  recipeActions,
  recipeSelectors,
  recipeReducer,
} from './recipe-slice';
