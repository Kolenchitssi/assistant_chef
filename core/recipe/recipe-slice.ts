import { createSlice /*, PayloadAction */ } from '@reduxjs/toolkit';

export interface IRecipeReducer {
  name: string;
  email: string;
  isFetching: boolean;
}

const initialState: IRecipeReducer = {
  name: '',
  email: '',
  isFetching: false,
};

export const recipeSlice = createSlice({
  name: 'accounts',
  initialState,
  reducers: {},
  selectors: {},
});

export const {
  actions: recipeActions,
  selectors: recipeSelectors,
  reducer: recipeReducer,
} = recipeSlice;
