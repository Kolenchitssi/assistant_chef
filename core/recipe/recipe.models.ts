export enum IngredientKeys {
  key = 'key',
  name = 'name',
  quantity = 'quantity',
  unitOfMeasurement = 'unitOfMeasurement',
}
export interface IIngredient {
  [IngredientKeys.key]: string;
  [IngredientKeys.name]: string;
  [IngredientKeys.quantity]: number;
  [IngredientKeys.unitOfMeasurement]: string;
}

export enum RecipeKeys {
  recipeID = 'recipeID',
  recipeName = 'Name',
  recipeDescription = 'description',
  ingredients = 'ingredients',
  mainIngredient = 'mainIngredient',
  imgUrl = 'imgUrl',
}

export interface IRecipe {
  [RecipeKeys.recipeID]?: string | null;
  [RecipeKeys.recipeName]: string;
  [RecipeKeys.recipeDescription]: string;
  [RecipeKeys.ingredients]: IIngredient[];
  [RecipeKeys.mainIngredient]: IIngredient;
  [RecipeKeys.imgUrl]: string | null;
}

export const DEFAULT_RECIPE = {
  [RecipeKeys.recipeID]: '',
  [RecipeKeys.recipeName]: '',
  [RecipeKeys.recipeDescription]: '',
  [RecipeKeys.ingredients]: [],
  [RecipeKeys.mainIngredient]: '',
};
