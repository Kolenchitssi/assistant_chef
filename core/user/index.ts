export * from './user.models';
export {
  userActions,
  userSelectors,
  userReducer,
} from './user-slice';
