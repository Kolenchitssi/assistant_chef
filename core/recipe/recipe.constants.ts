import { IngredientKeys } from './recipe.models';

export const INGREDIENT_INITIAL_VALUE = {
  [IngredientKeys.name]: '',
  [IngredientKeys.quantity]: 0,
  [IngredientKeys.unitOfMeasurement]: 'g',
};

export const UOM = ['g', 'kg', 'ml', 'l', 'pcs', 'tbsp', 'tsp'];
