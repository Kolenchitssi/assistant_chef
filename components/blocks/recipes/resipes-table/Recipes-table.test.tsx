import React from 'react';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { IRecipe } from '@/core/recipe';
import RecipesTable from './Recipes-table';

// Мок next/link
jest.mock('next/link', () => ({
  __esModule: true,
  default: ({
    children,
    href,
  }: {
    children: React.ReactNode;
    href: string;
  }) => <a href={href}>{children}</a>,
}));

// Мок компоненты @mantine/core
jest.mock('@mantine/core', () => {
  const ScrollContainer = ({
    children,
  }: {
    children: React.ReactNode;
  }) => <div>{children}</div>;
  const Thead = ({ children }: { children: React.ReactNode }) => (
    <thead>{children}</thead>
  );
  const Tbody = ({ children }: { children: React.ReactNode }) => (
    <tbody>{children}</tbody>
  );
  const Th = ({ children }: { children: React.ReactNode }) => (
    <th>{children}</th>
  );
  const Tr = ({ children }: { children: React.ReactNode }) => (
    <tr>{children}</tr>
  );
  const Td = ({ children }: { children: React.ReactNode }) => (
    <td>{children}</td>
  );

  const Table = ({
    children,
    striped,
    highlightOnHover,
  }: {
    children: React.ReactNode;
    striped?: boolean;
    highlightOnHover?: boolean;
  }) => (
    <table
      data-striped={striped}
      data-highlight-on-hover={highlightOnHover}
    >
      {children}
    </table>
  );

  Table.ScrollContainer = ScrollContainer;
  Table.Thead = Thead;
  Table.Tbody = Tbody;
  Table.Th = Th;
  Table.Tr = Tr;
  Table.Td = Td;

  const TextInput = ({
    value,
    onChange,
    placeholder,
  }: {
    value: string;
    onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
    placeholder: string;
  }) => (
    <input
      type="text"
      value={value}
      onChange={onChange}
      placeholder={placeholder}
    />
  );

  const Select = ({
    value,
    onChange,
    placeholder,
    data,
  }: {
    value: string | null;
    onChange: (value: string | null) => void;
    placeholder: string;
    data: string[];
  }) => (
    <select
      value={value || ''}
      onChange={(e) => onChange(e.target.value)}
    >
      <option value="" disabled>
        {placeholder}
      </option>
      {data.map((item) => (
        <option key={item} value={item}>
          {item}
        </option>
      ))}
    </select>
  );

  const MultiSelect = ({
    value,
    onChange,
    // placeholder,
    data,
  }: {
    value: string[];
    onChange: (value: string[]) => void;
    placeholder: string;
    data: string[];
  }) => (
    <select
      multiple
      value={value}
      onChange={(e) =>
        onChange(
          Array.from(
            e.target.selectedOptions,
            (option) => option.value,
          ),
        )
      }
    >
      {data.map((item) => (
        <option key={item} value={item}>
          {item}
        </option>
      ))}
    </select>
  );

  return {
    Table,
    TextInput,
    Select,
    MultiSelect,
  };
});

// Мок утилит
jest.mock('./resipes-table-utils', () => ({
  getUniqueIngredients: jest.fn(() => ['Ingredient1', 'Ingredient2']),
  getUniqueMainIngredients: jest.fn(() => [
    'MainIngredient1',
    'MainIngredient2',
  ]),
}));

// Моковые данные рецептов
const mockRecipes: IRecipe[] = [
  {
    recipeID: '1',
    Name: 'Test Recipe',
    description: 'Test Description',
    mainIngredient: {
      key: 'test-key',
      name: 'Test Main',
      quantity: 1,
      unitOfMeasurement: 'g',
    },
    ingredients: [],
    imgUrl: null,
  },
];

// describe для групировки тестов
describe('RecipesTable', () => {
  it('renders table headers', () => {
    render(<RecipesTable recipes={mockRecipes} />);
    //screen.debug(); // вывести весь DOM в конслоль для проверки

    // поиск текста на странице
    expect(screen.getByText('Name')).toBeInTheDocument();
    expect(screen.getByText('Main Ingredient')).toBeInTheDocument();
    expect(screen.getByText('Ingredients')).toBeInTheDocument();

    // если есть несколько на странице
    const descriptionElements = screen.getAllByText(/description/i); // Можно использовать регулярное выражение, здесь текст без учета регистра
    expect(descriptionElements.length).toBeGreaterThan(0); // убедиться, что есть совпадающие элементы
    // Или проверить конкретный элемент
    expect(descriptionElements[0]).toBeInTheDocument();

    // поиск элементов на странице
    expect(screen.getByRole('link')).toBeInTheDocument();
    expect(screen.getByRole('table')).toBeInTheDocument();

    // если проверяем элемент которого может не быть
    expect(screen.queryByRole('button')).toBeNull();
  });

  // test тоже самое что и it
  test('render empty list', () => {
    render(<RecipesTable recipes={[]} />);
    expect(screen.getByText('Name')).toBeInTheDocument();
  });

  // Проверка если при  изменении чего-то у нас изменилась структура таблицы
  test('table snapshot', () => {
    const table = render(<RecipesTable recipes={mockRecipes} />);
    expect(table).toMatchSnapshot();
  });

  // проверка действий  пользователя  ввод в поиск поля Name,
  // нужно дождатся обновления после ввода поэтому функция асинхронная
  test('Search Name is working', async () => {
    render(<RecipesTable recipes={mockRecipes} />);

    const user = userEvent.setup();

    await user.type(
      screen.getByPlaceholderText('Filter by name'),
      'myRecipe',
    );

    expect(screen.getByPlaceholderText('Filter by name')).toHaveValue(
      'myRecipe',
    );
  });
});
