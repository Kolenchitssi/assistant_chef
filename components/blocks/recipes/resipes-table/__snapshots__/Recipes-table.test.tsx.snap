// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipesTable table snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <table
          data-highlight-on-hover="true"
          data-striped="true"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                Description
              </th>
              <th>
                Main Ingredient
              </th>
              <th>
                Ingredients
              </th>
            </tr>
            <tr>
              <th>
                <input
                  placeholder="Filter by name"
                  type="text"
                  value=""
                />
              </th>
              <th>
                <input
                  placeholder="Filter by description"
                  type="text"
                  value=""
                />
              </th>
              <th>
                <select>
                  <option
                    disabled=""
                    value=""
                  >
                    Filter by ingredients
                  </option>
                  <option
                    value="MainIngredient1"
                  >
                    MainIngredient1
                  </option>
                  <option
                    value="MainIngredient2"
                  >
                    MainIngredient2
                  </option>
                </select>
              </th>
              <th>
                <select
                  multiple=""
                >
                  <option
                    value="Ingredient1"
                  >
                    Ingredient1
                  </option>
                  <option
                    value="Ingredient2"
                  >
                    Ingredient2
                  </option>
                </select>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <a
                  href="recipe/1"
                >
                  Test Recipe
                </a>
              </td>
              <td>
                Test Description
              </td>
              <td>
                <span>
                  Test Main
                  :
                  1
                  g
                </span>
              </td>
              <td />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <table
        data-highlight-on-hover="true"
        data-striped="true"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Description
            </th>
            <th>
              Main Ingredient
            </th>
            <th>
              Ingredients
            </th>
          </tr>
          <tr>
            <th>
              <input
                placeholder="Filter by name"
                type="text"
                value=""
              />
            </th>
            <th>
              <input
                placeholder="Filter by description"
                type="text"
                value=""
              />
            </th>
            <th>
              <select>
                <option
                  disabled=""
                  value=""
                >
                  Filter by ingredients
                </option>
                <option
                  value="MainIngredient1"
                >
                  MainIngredient1
                </option>
                <option
                  value="MainIngredient2"
                >
                  MainIngredient2
                </option>
              </select>
            </th>
            <th>
              <select
                multiple=""
              >
                <option
                  value="Ingredient1"
                >
                  Ingredient1
                </option>
                <option
                  value="Ingredient2"
                >
                  Ingredient2
                </option>
              </select>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <a
                href="recipe/1"
              >
                Test Recipe
              </a>
            </td>
            <td>
              Test Description
            </td>
            <td>
              <span>
                Test Main
                :
                1
                g
              </span>
            </td>
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
